### PR TITLE
Core-1249  Elasticsearch Drain Utility

### PR DESCRIPTION
**Pull request by athompson-r7**
The purpose of this utility is to drain documents from Elasticsearch nodes in an AutoScaling Group.

This will help you do Elasticsearch node replacement while keeping the cluster healthy.

### CODE REVIEW

**Comment by rvanburen-r7** [comment]
Looks good to me, but you may want a second pair of eyes.

File: Rakefile [hunk 1]
```diff
@@ -1,2 +1,68 @@
...
+
+def elasticsearch_command
+  path = 'tmp/elasticsearch-1.7.2/bin/elasticsearch'
+  path = ::File.expand_path(path, __dir__)
+  "bash #{path}"
+end
+
+ENV['TEST_CLUSTER_NODES'] = '1'
+ENV['TEST_CLUSTER_COMMAND'] = elasticsearch_command
+
```
Comment (fmitchell-r7) [inline Rakefile]: Maybe expose ES version as an environment variable too?

File: README.md [hunk 2]
```diff
@@ -1,30 +1,30 @@
...
+# Elasticsearch Drain
 
-TODO: Write a gem description
+The purpose of this utility is to drain documents from Elasticsearch nodes in an AutoScaling Group.
 
-## Installation
+This will help you do Elasticsearch node replacement while keeping the cluster healthy.
 
-Add this line to your application's Gemfile:
+Consider the following deployment procedure:
```
Comment (fmitchell-r7) [inline README.md]: Might want to include a note that this is a useful way to do an upgrade if you can't update binaries in place, like when you bake versioned AMIs.

File: lib/elasticsearch/drain.rb [hunk 3]
```diff
@@ -1,7 +1,113 @@
...
+        hosts: hosts,
+        retry_on_failure: true,
+        log: true,
+        logger: ::Logger.new('es_client.log', 10, 1_024_000)
+      )
+    end
+
+    # EC2 AutoScaling Client
+    def asg
+      AutoScaling.new(@asg_name, @region)
```
Comment (fmitchell-r7) [inline lib/elasticsearch/drain.rb]: Worth memoizing like `client` does?

~~~ruby
@asg ||= AutoScaling.new(@asg_name, @region)
~~~

File: lib/elasticsearch/drain.rb [hunk 4]
```diff
@@ -1,7 +1,113 @@
...
+    # @return [Array<OpenStruct>] Array of node objects
+    def nodes
+      Nodes.new(client, asg)
+    end
+
+    # Convience method to access {Elasticsearch::Drain::Cluster#cluster}
+    #
+    # @return [Elasticsearch::API::Cluster] Elasticsearch cluster client
+    def cluster
+      Cluster.new(client)
```
Comment (fmitchell-r7) [inline lib/elasticsearch/drain.rb]: Worth memoizing?
~~~ruby
@cluster_client ||= Cluster.new(client)
~~~

File: lib/elasticsearch/drain.rb [hunk 5]
```diff
@@ -1,7 +1,113 @@
...
+
+    def active_nodes_in_asg
+      instances = asg.instances
+      nodes.nodes_in_asg(reload: true, instances: instances)
+    end
+
+    def drain
+      fail 'Cluster is unhealthy' unless cluster.healthy?
+      instances = asg.instances
+      nodes_to_drain = nodes.nodes_in_asg(reload: true, instances: instances)
```
Comment (fmitchell-r7) [inline lib/elasticsearch/drain.rb]: Can you call `self.active_nodes_in_asg` here instead?

File: README.md (hunk 2 above)
Comment (athompson-r7) [inline README.md]: Addressed in 49ac49c.

**Comment by athompson-r7** [comment]
@fmitchell-r7 I'm still looking at one last issue with memoizing the asg client; in the mean time can you take a look at the recent changes.  I refactored the cli interface, like I should have in the first place.  I moved things around so that I could use the thor magic. 

https://github.com/rapid7/elasticsearch-drain/compare/26198728790ed6425ce5028051506728cc43c0f5...CORE-1249

**Comment by fmitchell-r7** [comment]
:+1:

**Pull request merged**